### ISSUE TITLE
TripletSemiHardLoss produces only NaNs when training with a TPU

### ISSUE THREAD

**Issue by kbelenky** [keras-team/keras#15842]
This bug was discovered in Google Colab, but appears to be rooted in TripletSemiHardLoss.

Here's an example Colab that illustrates the problem:
https://colab.research.google.com/drive/10GuYAFDMuxXe2jqdj7hy_I7XcApJFmj7?usp=sharing

I combined the example Colab here:
https://colab.research.google.com/github/GoogleCloudPlatform/training-data-analyst/blob/master/courses/fast-and-lean-data-science/01_MNIST_TPU_Keras.ipynb

with some of the steps here:
https://www.tensorflow.org/addons/tutorials/losses_triplet

Specifically:
I deleted the line: label = `tf.one_hot(label, 10)`
I replaced the last layer of the network (Dense ReLU) with these two layers:

```
 tf.keras.layers.Dense(10, activation=None),
 tf.keras.layers.Lambda(lambda x: tf.math.l2_normalize(x, axis=1)) # L2 normalize embeddings
```

And then changed the loss function in make_model to: `tfa.losses.TripletSemiHardLoss(distance_metric='angular', margin=0.2)`

When run on a CPU or GPU, the training works as expected and the loss function produces reasonable results.
When run on a TPU, the training runs (as in it does not produce an error), but the loss function produces only NaN, causing the training to effectively fail.

I understand that not all operations are available on TPUs and TripletSemiHardLoss may fundamentally not work on a TPU, but it should at least fail with an informative error message.

**Comment by kumariko** [comment]
@sachinprasadhs Was able to reproduce the issue in TF v2.7.0 ,please find the [gist](https://colab.research.google.com/gist/kumariko/69901db667512210a0f3f9d2e721f628/tripletsemihardloss-fails-on-tpu-but-not-cpu-gpu.ipynb#scrollTo=KDK8BIYTjelZ) here.Thanks!

**Comment by qlzh727** [comment]
Shouldn't this bug be filed to tf.addon repository?

**Comment by Shiro-LK** [comment]
Does anyone know why it is not working with TPU only ? 
Is it during the distribution of the batch for each core ? Or is it some issue with specific function inside the loss ?

**Comment by tilakrayal** [comment]
Hello, Thank you for reporting an issue.

We're currently in the process of migrating the new `Keras 3` code base from **keras-team/keras-core** to **keras-team/keras.**
Consequently, This issue may not be relevant to the `Keras 3 `code base. After the migration is successfully completed, feel free to reopen this issue at **keras-team/keras** if you believe it remains relevant to the `Keras 3` code base. 
If instead this issue is a bug or security issue in legacy **tf.keras**, you can instead report a new issue at [keras-team/tf-keras](https://github.com/keras-team/tf-keras/issues), which hosts the TensorFlow-only, legacy version of Keras.

To know more about `Keras 3,` please take a look at https://keras.io/keras_core/announcement/. Thank you!
